katsdpcontroller/product_controller.py

```diff
@@ -907,6 +899,43 @@ async def gain_impl(
         """
         raise NotImplementedError()
 
+    async def gain_all(self, stream_name: str, values: Sequence[str]) -> None:
+        """Set F-engine gains for all inputs."""
+        if self.state not in {ProductState.CAPTURING, ProductState.IDLE}:
+            raise FailReply(f"Cannot set gains in state {self.state}")
+        stream = self._find_stream(stream_name)
+        if not isinstance(stream, product_config.GpucbfAntennaChannelisedVoltageStream):
+            raise FailReply(f"Stream {stream_name!r} is of the wrong type")
+        if len(values) not in {1, stream.n_chans}:
+            raise FailReply(f"Expected 0, 1, or {stream.n_chans} values, received {len(values)}")
+        values = tuple(values)
+        if values != ("default",):
+            try:
+                for v in values:
+                    complex(v)  # Evaluated just for the exception
+            except ValueError as exc:
+                raise FailReply(str(exc))
+        # TODO: remove the type: ignores once SDPSubarrayProduct and SDPSubarrayProductBase merged
+        await self._multi_request(  # type: ignore
+            self.find_nodes(task_type="f", streams=[stream]),  # type: ignore
+            itertools.repeat(("gain-all",) + values)
```

Comment:
Do the f-engines need to know about the gain-all request to the PC? My understanding was that the PC would just translate the gain-all request to individual (identical) gain requests for each F-engine. Or maybe I'm misunderstanding.

```suggestion
            itertools.repeat(("gain",) + values)
```

Edit: I think I've just understood, in that `?gain` needs the name of the input as well, which in the case of the F-engine will (I guess) correspond to the polarisation being adjusted.

So my suggestion won't work - either we'd need to know the names of the inputs that each F-engine has, or to implement a name-agnostic `?gain-all` at the F-engine which I think was your original intention.

Reply:
It could issue two `gain` commands (one for each pol) without too much trouble, but that would double the network bandwidth, wouldn't work so well with `_multi_request`, and it wouldn't handle the magic `default` value. I figure it will be better to just implement `gain-all` within fgpu.